File: computer3/Fall Term/Solutions/Class09_Nov_06/lvl3_directed_acyclic_graph.cpp
#include <bits/stdc++.h>
using namespace std;
int n, adj[1002][1002], vis[1002]; bool cycle;
void dfs(int u){
    vis[u] = 1;
    for(int v=1; v<=n; v++){
        if(adj[u][v] == 1){
            if(vis[v] == 0) dfs(v);
            else if(vis[v] == 1) cycle=true;
        }
    }
    vis[u] = 2;
}
int main(){
    cin >> n;
    for(int i=1; i<=n; i++)
        for(int j=1; j<=n; j++)
            cin >> adj[i][j];
    for(int i=1; i<=n; i++)
        if(!vis[i]) dfs(i);
    cout << (cycle? "NO": "YES") << endl;
}

File: computer3/Fall Term/Solutions/Class09_Nov_06/lvl3_directed_acyclic_graph.java
import java.util.*;
import java.io.*;
public class homework {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static int n, vis[], adj[][]; static boolean cycle = false;
	public static void main(String[] args) throws IOException{
		n = readInt(); vis = new int[n]; adj = new int[n][n];
		for(int i=0; i<n; i++)
			for(int j=0; j<n; j++)
				adj[i][j] = readInt();
		for(int i=0; i<n; i++) {
			if(vis[i] == 0) dfs(i);
		}
		System.out.println(cycle? "NO":"YES");
	}
	static void dfs(int u) {
		vis[u] = 1;
		for(int v=0; v<n; v++) {
			if(adj[u][v] == 1) {
				if(vis[v] == 1) { cycle = true; return; }
				else if (vis[v] == 0) dfs(v);
			}
		}
		vis[u] = 2;
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}